gtk4-macros/src/lib.rs

```diff
@@ -90,3 +94,147 @@ pub fn composite_template_derive(input: TokenStream) -> TokenStream {
+///     first value and can take self again. Mainly useful for callbacks that are invoked with
+///     swapped="true".
+///
+/// The `rest` attribute can be placed on the last argument of a template callback. This attribute
+/// must be used on an argument of type `&[glib::Value]` and will pass in the remaining arguments.
+/// The first and last values will be omitted from the slice if this callback is a `function`.
+///
+/// # Example
+///
+/// ```no_run
+/// # fn main() {}
+/// use gtk::prelude::*;
+/// use gtk::glib;
+/// use gtk::CompositeTemplate;
+/// use gtk::subclass::prelude::*;
+///
+/// mod imp {
+///     use super::*;
+///
+///     #[derive(Debug, Default, CompositeTemplate)]
+///     #[template(file = "test/template_callbacks.ui")]
+///     pub struct MyWidget {
+///         #[template_child]
+///         pub label: TemplateChild<gtk::Label>,
+///         #[template_child(id = "my_button_id")]
+///         pub button: TemplateChild<gtk::Button>,
+///     }
+///
+///     #[glib::object_subclass]
+///     impl ObjectSubclass for MyWidget {
+///         const NAME: &'static str = "MyWidget";
+///         type Type = super::MyWidget;
+///         type ParentType = gtk::Box;
+///
+///         fn class_init(klass: &mut Self::Class) {
+///             Self::bind_template(klass);
+///             // Bind the private callbacks
+///             Self::bind_template_callbacks(klass);
+///             // Bind the public callbacks
+///             Self::Type::bind_template_callbacks(klass);
+///             // Bind callbacks from another struct
+///             super::Utility::bind_template_callbacks(klass);
+///         }
+///
+///         fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+///             obj.init_template();
+///         }
+///     }
+///
+///     #[gtk::template_callbacks(private)]
+///     impl MyWidget {
+///         #[template_callback]
+///         fn button_clicked(&self, button: gtk::Button) {
+///             button.set_label("I was clicked!");
+///             self.label.set_label("The button was clicked!");
+///         }
+///         #[template_callback(function, name = "strlen")]
+///         fn string_length(s: &str) -> u64 {
+///             s.len() as u64
+///         }
+///     }
+///
+///     impl ObjectImpl for MyWidget {}
+///     impl WidgetImpl for MyWidget {}
+///     impl BoxImpl for MyWidget {}
+/// }
+///
+/// glib::wrapper! {
+///     pub struct MyWidget(ObjectSubclass<imp::MyWidget>) @extends gtk::Widget, gtk::Box;
+/// }
+///
+/// #[gtk::template_callbacks]
+/// impl MyWidget {
+///     pub fn new() -> Self {
+///         glib::Object::new(&[]).expect("Failed to create an instance of MyWidget")
+///     }
+///     #[template_callback]
+///     pub fn print_both_labels(&self) {
+///         let self_ = imp::MyWidget::from_instance(self);
+///         println!("{} {}", self_.label.label(), self_.button.label().unwrap().as_str());
+///     }
+/// }
```

Comment:
Is this use case useful at all? Do you ever need to have the template callbacks defined on the public type's impl?

Reply:
I think it could be useful, for signal handlers that can also be called as public methods.